### PR TITLE
Lazy2d

### LINKED ISSUE #14: A "lazy" 2D rotation type
A 2D rotation parameterized by an angle would be welcome. This would go in the "angle-axis" family.

Internally the new type could store the angle, or it could optionally also store the cos and sin of the angle for efficiency in application to multiple points. Would only save one value compared to `RotMatrix{2}` and be marginally faster, but it would be guaranteed to be a parity-preserving rotation.

### PR DESCRIPTION
**Pull request by jw3126**
fix #14. 
I noticed one gotcha (that is also present in `AngleAxis`).
```julia
julia> using Rotations

julia> r = Angle2d(1); @show eltype(r); @show typeof(r[1])
eltype(r) = Int64
typeof(r[1]) = Float64

```

### CODE REVIEW

File: src/core_types.jl [hunk 1]
```diff
@@ -125,6 +125,55 @@ Base.inv(r::RotMatrix) = RotMatrix(r.mat')
...
+        throw(DimensionMismatch("Cannot rotate a vector of length $(length(v))"))
+    end
+    x,y = v
+    s,c = sincos(r.theta)
+    T = eltype(r)
+    similar_type(v,T)(c*x - s*y, s*x + c*y)
+end
+
+Base.:*(r1::Angle2d, r2::Angle2d) = Angle2d(r1.theta + r2.theta)
+Base.:^(r::Angle2d, t::Real) = Angle2d(r.theta^t)
```
Comment (jw3126) [inline src/core_types.jl]: This is wrong.

File: src/core_types.jl [hunk 2]
```diff
@@ -125,6 +125,55 @@ Base.inv(r::RotMatrix) = RotMatrix(r.mat')
...
+    s,c = sincos(r.theta)
+    T = eltype(r)
+    similar_type(v,T)(c*x - s*y, s*x + c*y)
+end
+
+Base.:*(r1::Angle2d, r2::Angle2d) = Angle2d(r1.theta + r2.theta)
+Base.:^(r::Angle2d, t::Real) = Angle2d(r.theta*t)
+Base.:^(r::Angle2d, t::Integer) = Angle2d(r.theta*t)
+Base.inv(r::Angle2d) = Angle2d(-r.theta)
+Base.transpose(r::Angle2d) = inv(r)
```
Comment (andyferris) [inline src/core_types.jl]: This method is already defined in core_types.jl

File: src/core_types.jl [hunk 3]
```diff
@@ -125,6 +125,55 @@ Base.inv(r::RotMatrix) = RotMatrix(r.mat')
...
+rotation_angle(r::Angle2d) = r.theta
+
+@inline function Base.getindex(r::Angle2d, i::Int)
+    if i == 1
+        cos(r.theta)
+    elseif i ==2
```
Comment (andyferris) [inline src/core_types.jl]: Spacing issue

**Comment by andyferris** [comment]
> I noticed one gotcha (that is also present in AngleAxis)

Yes, all of these structs need two parameters, unfortunately (one for the storage type and one for the computed element type).

**Comment by jw3126** [comment]
@andyferris merge?

**Comment by colinxs** [comment]
I'd love if you could merge this as well. I'm currently using a fork in a project that implements similar functionality, but would like to switch back to upstream for distribution.

**Comment by tkoolen** [comment]
Looks good to me, so I'm going to go ahead and merge this.

**Pull request merged**